### PR TITLE
feat: [CI-4646]: add github oauth type support

### PR DESCRIPTION
**Pull request by HenAmar**
## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33948)

### CODE REVIEW

**Comment by aman-harness** [comment]
Great work Hen. Adding @harshjain12 also for review so we don't miss any corner cases on the connector side.

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/github/GithubOauthDTO.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+@Data
+@Builder
+@JsonIgnoreProperties({"userName"})
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@ApiModel("GithubOauthDTO")
+@Schema(name = "GithubOauth",
+        description = "This contains details of the Github credentials Specs for oauth")
+public class GithubOauthDTO implements GithubHttpCredentialsSpecDTO {
+    public static String userName = "Oauth";
```
Comment (harshjain12) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/github/GithubOauthDTO.java:25]: do we have to call Oauth ?

**Comment by harshjain12** [comment]
@HenAmar It is like a new flavour of git connector.

We have to think and test following with new change

- Supporting Git clone via drone plugin as plugin takes user name .  
- Calling git apis via token [LATEST_COMMIT_ID, PULL_REQUEST_WITH_COMMITS]
- Check whether we have to support git sync with this;
-  Support git commit status with this flavour of connector
-  Perpetual task to check connectivity.

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/github/GithubOauthDTO.java (hunk 1 above)
Comment (HenAmar) [reply 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/scm/github/GithubOauthDTO.java]: No, not mandatory but the user will not be aware of it.

**Comment by HenAmar** [comment]
trigger ti0

**Comment by HenAmar** [comment]
trigger authorcheck

**Review by harshjain12 — APPROVED** [review]
> Reviewed 2 of 10 files at r1, 9 of 12 files at r2, 4 of 5 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33948)* status: :shipit: complete! ✓

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 2 of 10 files at r1, 9 of 12 files at r2, 4 of 5 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33948)* status: :shipit: complete! ✓

**Pull request merged**